packages/pds/src/db/util.ts

```diff
@@ -19,22 +21,87 @@ export const actorWhereClause = (actor: string) => {
 
 export const countAll = sql<number>`count(*)`
 
-export const paginate = <QB extends SelectQueryBuilder<any, any, any>>(
+export const paginate = <QB extends SelectQueryBuilder<any, any, any>, T>(
   qb: QB,
-  opts: {
-    limit?: number
-    before?: string
-    by: DbRef
-    secondaryOrder?: DbRef
-  },
+  opts:
+    | {
+        limit?: number
+        before?: string
+        by: DbRef
+      }
+    | {
+        limit?: number
+        before?: string
+        keyset: Keyset<T>
+      },
 ) => {
+  if ('by' in opts) {
+    return qb
+      .if(opts.limit !== undefined, (q) => q.limit(opts.limit as number))
+      .orderBy(opts.by, 'desc')
+      .if(opts.before !== undefined, (q) => q.where(opts.by, '<', opts.before))
+  }
+  const cursor = opts.keyset.unpack(opts.before)
+  const keysetCondition = getKeysetCondition(cursor, opts.keyset)
   return qb
-    .orderBy(opts.by, 'desc')
-    .if(opts.secondaryOrder !== undefined, (q) =>
-      q.orderBy(opts.secondaryOrder as DbRef, 'desc'),
-    )
     .if(opts.limit !== undefined, (q) => q.limit(opts.limit as number))
-    .if(opts.before !== undefined, (q) => q.where(opts.by, '<', opts.before))
+    .orderBy(opts.keyset.primary, 'desc')
+    .orderBy(opts.keyset.secondary, 'desc')
+    .if(keysetCondition !== undefined, (qb) =>
+      keysetCondition ? qb.where(keysetCondition) : qb,
+    )
+}
+
+type DefaultRow = { createdAt: string; uri: string }
+export abstract class Keyset<T = DefaultRow> {
+  abstract primary: DbRef
+  abstract secondary: DbRef
+  cursorFromResult(result: T): Cursor
+  cursorFromResult<T extends DefaultRow>(result: T): Cursor {
+    return {
+      primary: result.createdAt,
+      secondary: result.uri,
+    }
+  }
+  packFromResult(results: T | T[]): string | undefined {
+    const result = Array.isArray(results) ? results.at(-1) : results
+    if (result === undefined) return
+    return this.pack(this.cursorFromResult(result))
+  }
+  pack(cursor?: Cursor): string | undefined {
+    if (cursor === undefined) return
+    return JSON.stringify([cursor.primary, cursor.secondary])
```

Comment:
hmm maybe strange for the JSON schema to say this type is "string" when it's stringified JSON. Wondering if it may make more sense to separate it with some delimiter 🧐

Reply:
My thinking is that:
 - For consumers the cursor is an opaque value, they shouldn't really muck around in there or try to build one themselves.
 - I reached for a json tuple just because it's a foolproof way to stick two things together and then take them apart without worrying about what characters are allowed in each part.

Admittedly it's a bit ugly-looking, but it has the properties I was going for.  Definitely not married to it!